ehrql/backends/tpp.py

```diff
@@ -79,8 +79,8 @@ def modify_query_variables(self, variables):
             SELECT
                 addr.Patient_ID AS patient_id,
                 addr.PatientAddress_ID AS address_id,
-                CAST(addr.StartDate AS date) AS start_date,
-                CAST(addr.EndDate AS date) AS end_date,
+                CAST(NULLIF(addr.StartDate, '9999-12-31') AS date) AS start_date,
+                CAST(NULLIF(addr.EndDate, '9999-12-31') AS date) AS end_date,
```

Comment:
I think that for a couple of subtle, non-obvious reasons this might not be quite right.

One is that, before we do the casting, `StartDate` and `EndDate` are actually datetimes not dates and so we should really be using a datetime literal to compare them with like we do here:
https://github.com/opensafely-core/ehrql/blob/aeec978ee040f57cded5fc9bb804f5020473cafc/ehrql/backends/tpp.py#L173

In practice, MSSQL will do an implicit cast of the date to a datetime so the comparison won't throw an error. But it would be better not to rely on that behaviour.

The second issue is that `YYYY-MM-DD` isn't a reliable date format in MSSQL because, depending on the localisation settings, it can actually interpret it as `YYYY-DD-MM`. The reliable date format for MSSQL is the non-standard `YYYYMMDD` format (i.e. `99991231`). It's kind of incredible that this is the case, and it's not the sort of thing I'd expect many people to know. Of course, this goes away if we switch to using datetimes (the reliable format for datetimes _does_ have dashes in).

```suggestion
                CAST(NULLIF(addr.StartDate, '9999-12-31T00:00:00') AS date) AS start_date,
                CAST(NULLIF(addr.EndDate, '9999-12-31T00:00:00') AS date) AS end_date,
```

Reply:
😱 Thanks for highlighting this, Dave. (And also for your description of the issue in #92.)

Reply:
Thanks Dave! And it looks like we did do this before but I missed it, e.g., https://github.com/opensafely-core/ehrql/blob/aeec978ee040f57cded5fc9bb804f5020473cafc/ehrql/backends/tpp.py#L173

But looking through the code I found another datetime variable (EndDate from [RegistrationHistory](https://reports.opensafely.org/reports/opensafely-tpp-database-schema/#RegistrationHistory)) where we currently just check for the date in YYYY-MM-DD format:

https://github.com/opensafely-core/ehrql/blob/aeec978ee040f57cded5fc9bb804f5020473cafc/ehrql/backends/tpp.py#L549

Is this something where we should switch to `9999-12-31T00:00:00` as well"?

Reply:
Already done! :grinning: 
https://github.com/opensafely-core/ehrql/pull/1708

Reply:
hahahahahah, so fast!